pages/lessons/fundamentals/wolovim-part-1.mdx

```diff
@@ -0,0 +1,321 @@
+
+### **Installation**
+
+In this walkthrough, we’ll just be working within a Python interpreter. We won't be creating any directories, files, classes or functions.
+
+<Callout emoji='💡' size='md' variant='info'>
+**Note:** In the examples below, commands that begin with `$` are intended to be run in the terminal. (Do not type the `$`, it just signifies the start of the line.)
+</Callout>
+
+First, install [IPython](https://ipython.org/) for a user-friendly environment to explore in. IPython offers tab completion, among other features, making it much easier to see what’s possible within web3.py.
+
+```bash
+$ pip install ipython
+```
+
+web3.py is published under the name `web3`. Install the latest version like so:
+
+```bash
+$ pip install web3
+```
+
+One more thing – we're going to simulate a blockchain later, which requires a couple more dependencies. You can install those via:
+
+```bash
+$ pip install "web3[tester]"
+```
+
+You’re all set up to go!
+
+### **Spin up a sandbox**
+
+Open up a new Python environment by running `ipython` in your terminal. This is comparable to running `python`, but comes with more bells and whistles.
+
+```bash
+$ ipython
+```
+
+This will print out some information about the versions of Python and IPython you’re running, then you should see a prompt waiting for input:
+
+```python
+In [1]:
+```
+
+You’re looking at an interactive Python shell now. Essentially, its a sandbox to play in. If you’ve made it this far, its time to import web3.py:
+
+```python
+In [1]: from web3 import Web3
+```
+
+### **Introducing the Web3 module**
+
+Besides being a gateway to Ethereum, the [Web3](https://web3py.readthedocs.io/en/stable/overview.html#base-api) module offers a few convenience functions. Let’s explore a couple.
+
+In an Ethereum application, you will commonly need to convert currency denominations. The Web3 module provides a couple of helper methods just for this: [from_wei](https://web3py.readthedocs.io/en/stable/web3.main.html#web3.Web3.fromWei) and [to_wei](https://web3py.readthedocs.io/en/stable/web3.main.html#web3.Web3.toWei).
+
+<Callout emoji='💡' size='md' variant='info'>
+**Note:** Computers are notoriously bad at handling decimal math. To get around this, developers often store dollar amounts in cents. For example, an item with a price of $5.99 may be stored in the database as 599. A similar pattern is used when handling transactions in ether. However, instead of two decimal points, ether has 18! The smallest denomination of ether is called *wei*, so that’s the value specified when sending transactions.
+
+1 ether = 1000000000000000000 wei
+
+1 wei = 0.000000000000000001 ether
+</Callout>
+
+Try converting some values to and from wei. Note that there are [names for many of the denominations](https://web3py.readthedocs.io/en/stable/examples.html#converting-currency-denominations) in between ether and wei. One of the better known among them is **gwei**, as it’s often how transaction fees are represented.
+
```

Comment:
Hey Marc, 
I was getting a couple of issues:
1. I noticed the syntax and also prompt that the terminal is offering me e.g. `Web3.to_wei(1, 'ether')`, and by trial and error, I got it to accept `....toWei`, which led me to believe there was a version clash.  Not sure what I did wrong, but you'll see from the images I dropped in Discord that I ended up with an older version of web3.py i.e. Version: 5.31.3. I'm going to retrace my steps tomorrow. 

I have to be honest, as a rookie, I had to figure out how to do a lot of things myself based on what there is and isn't in the lesson content.

I'll start afresh tomorrow with a new head on. And see where my attention maybe got distracted.

Reply:
on this point, you've definitely got an outdated web3.py version. the major version bump to v6 will resolve the `to_wei` issue. can only speculate as to how you've got a v5 version, but `pip install web3 --upgrade` should do the trick. wondering if i should lock the version in the lesson.

Reply:
maybe this is a doc/reference we should be working off @wolovim ?
https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/#creating-a-virtual-environment
It's the only doc I seem to be coming across that actually uses the `python3` (as opposed to `python`) executable.

Reply:
Hey @wolovim 

I was just testing myself to see how quickly I could do a quick guide, and it's probably full of errors, and then there's the version control issue, but some things along these lines might be really helpful as a side drawer just before the *Installation* section. Would you be able to throw something like this together? It would really make a great lesson accessible to all levels, otherwise it's a real *cufuffle* trying to figure out what goes where:

How a snake charmer would set up a virtual environment for Python. 
Let us see what version of python you have installed already.
Open a terminal and run
`bash python`
If that gives you no result, that’s actually good as you don’t have python2 or lower installed, now run:
` bash python3 `
If you have python 3.7 or greater, you’re ready to set up your virtual environment, if not, download the latest version of (python3)[]

Now, let’s create a directory to host your venv. I’m calling mine `venv_eth_part1`
`bash // create project directory:
mkdir venv_eth_part1

// navigate into directory
cd venv_eth_part1`

Create the virtual environment:
`bash // unix/macOS:
python3 -m venv env 			

// windows:
py -m venv env`
           
Now activate the environment:
`bash // unix/macOS:
source env/bin/activate		

// windows:
.\env\Scripts\activate`

Double check what your current working directory is:
`bash // unix/macOS:
which python	
// output should be:
.../env/bin/python			

// windows:
where python
// output should be:
...\env\Scripts\python.exe`

How do you deactivate your virtual environment? Easy :) :
`// unix/macOS:
deactivate					

//windows:
deactivate`

If and when you need to use a virtual environment again with all the same dependencies, navigate back to your directory and simply run:
`bash // unix/macOS:
source env/bin/activate		

// windows:
.\env\Scripts\activate`

Now you are ready to set up your dependencies inside your new safe environment, so head back to Installation 👈🏼